### PR TITLE
Add support for dataloader_iter to validate and test steps

### PR DESCRIPTION
**Pull request by ananthsub**
## What does this PR do?

This extends experimental support for accessing the `dataloader_iter` in the validation and test steps for customized inter-batch parallelism support

### Does your PR introduce any breaking changes? If yes, please list them.
None

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py [hunk 1]
```diff
@@ -105,9 +105,13 @@ def advance(  # type: ignore[override]
...
 
         assert self._dataloader_iter is not None
-        batch, self.batch_progress.is_last_batch = next(self._dataloader_iter)
-        if batch is None:
-            raise StopIteration
+        if not isinstance(data_fetcher, DataLoaderIterDataFetcher):
+            batch_idx = self.batch_idx + 1
+            batch, self.batch_progress.is_last_batch = next(self._dataloader_iter)
+            if batch is None:
+                raise StopIteration
```
Comment (ananthsub) [inline pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:112]: @carmocca @awaelchli i wasn't sure why this raises a stop iteration in eval but not in training:
https://github.com/PyTorchLightning/pytorch-lightning/blob/f41d1e5e5ebb7040a39d137695e818cada9a9234/pytorch_lightning/loops/epoch/training_epoch_loop.py#L162-L164

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py [hunk 2]
```diff
@@ -105,9 +105,12 @@ def advance(  # type: ignore[override]
         void(dl_max_batches)
 
         assert self._dataloader_iter is not None
-        batch, self.batch_progress.is_last_batch = next(self._dataloader_iter)
-        if batch is None:
-            raise StopIteration
+        if not isinstance(data_fetcher, DataLoaderIterDataFetcher):
+            batch, self.batch_progress.is_last_batch = next(self._dataloader_iter)
```
Comment (tchaton) [inline pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:109]: It would be better to track the batch_idx from the DataFetcher automatically as I did with the DataLoaderIterDataFetcher

File: pytorch_lightning/trainer/connectors/data_connector.py [hunk 3]
```diff
@@ -127,16 +127,25 @@ def on_trainer_init(
         self.trainer._is_data_prepared = False
 
     def _select_data_fetcher(self) -> AbstractDataFetcher:
+        running_stage_to_step_method_name = {
+            RunningStage.TRAINING: "training_step",
+            RunningStage.SANITY_CHECKING: "validation_step",
+            RunningStage.VALIDATING: "validation_step",
```
Comment (tchaton) [inline pytorch_lightning/trainer/connectors/data_connector.py:133]: Should we include predict ?

File: tests/utilities/test_fetching.py [hunk 4]
```diff
@@ -262,6 +262,56 @@ def training_epoch_end(self, *_):
...
+            batch = next(dataloader_iter)
+            return super().validation_step(batch, batch_idx)
+
+        def test_step(self, dataloader_iter, batch_idx):
+            batch = next(dataloader_iter)
+            return super().test_step(batch, batch_idx)
+
+    model = TestModel()
+    trainer = Trainer(default_root_dir=tmpdir, max_epochs=1)
+    if fn == "fit":
```
Comment (tchaton) [inline tests/utilities/test_fetching.py:307]: I don't believe you need to test training_step as done previously.

File: pytorch_lightning/trainer/connectors/data_connector.py (hunk 3 above)
Comment (ananthsub) [reply pytorch_lightning/trainer/connectors/data_connector.py:133]: Good point. I want to add support to predict in a separate PR to minimize the changes here

Comment (tchaton) [reply pytorch_lightning/trainer/connectors/data_connector.py:133]: Good :)

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:109]: @ananthsub Can we prevent this check. This would make the code less readable and this was already supported in the past. I don't believe the `training_step` API should change. Batch would just be renamed dataloader_iter, but the batch_idx should still be the exact number of batches fetched by the user.

Comment (tchaton) [reply pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:109]: As done there: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/utilities/fetching.py#L326

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:111]: I believe it is not necessary as the fetchers will take care of it. You can try removing it and see if CI passes.

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:109]: What Ananth is adding here matches exactly what we have in the `TrainingEpochLoop`

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/loops/dataloader/evaluation_loop.py [hunk 5]
```diff
@@ -110,7 +120,8 @@ def on_run_start(self, *args: Any, **kwargs: Any) -> None:
         hooks."""
         void(*args, **kwargs)
 
-        self._data_fetcher = DataFetcher()
+        data_fetcher_cls = _select_data_fetcher_type(self.trainer)
+        self._data_fetcher = data_fetcher_cls()
```
Comment (ananthsub) [inline pytorch_lightning/loops/dataloader/evaluation_loop.py:124]: @carmocca - n00b question: why return the type vs the instance in #11621 
 ?

here i'm simply copy/pasting that style but was wondering if you had a specific reason its done this way: 

https://github.com/PyTorchLightning/pytorch-lightning/blob/f79d75d4b5e50eb3e37073606eba5ca10d48c99a/pytorch_lightning/loops/fit_loop.py#L346-L358

Comment (carmocca) [reply pytorch_lightning/loops/dataloader/evaluation_loop.py:124]: In case one wants to modify the number of prefetch batches passed.

File: pytorch_lightning/loops/dataloader/evaluation_loop.py [hunk 6]
```diff
@@ -366,3 +377,21 @@ def _print_results(results: List[_OUT_DICT], stage: str, file: Optional[IO[str]]
...
+
+
+def _select_data_fetcher_type(trainer: "pl.Trainer") -> Type[AbstractDataFetcher]:
+    lightning_module = trainer.lightning_module
+    step_fx = (
+        getattr(lightning_module, "test_step") if trainer.testing else getattr(lightning_module, "validation_step")
+    )
+    if is_param_in_hook_signature(step_fx, "dataloader_iter", explicit=True):
+        rank_zero_warn(
+            "Found `dataloader_iter` argument in the `training_step`. Note that the support for "
```
Comment (carmocca) [inline pytorch_lightning/loops/dataloader/evaluation_loop.py:389]:
```suggestion
            f"Found `dataloader_iter` argument in the `{step_fx_name}`. Note that the support for "
```

File: pytorch_lightning/loops/dataloader/evaluation_loop.py [hunk 7]
```diff
@@ -366,3 +377,21 @@ def _print_results(results: List[_OUT_DICT], stage: str, file: Optional[IO[str]]
...
+            "this signature is experimental and the behavior is subject to change."
+        )
+        return DataLoaderIterDataFetcher
+    elif os.getenv("PL_INTER_BATCH_PARALLELISM", "0") == "1":
+        if not isinstance(trainer.accelerator, GPUAccelerator):
+            raise MisconfigurationException("Inter batch parallelism is available only when using NVidia GPUs.")
```
Comment (carmocca) [inline pytorch_lightning/loops/dataloader/evaluation_loop.py:395]: nit: let's not copy over this typo

```suggestion
            raise MisconfigurationException("Inter batch parallelism is available only when using Nvidia GPUs.")
```

File: tests/utilities/test_fetching.py [hunk 8]
```diff
@@ -263,6 +263,56 @@ def training_epoch_end(self, *_):
...
+
+            batch = self.batches.pop(0)
+            assert isinstance(batch, torch.Tensor) or batch is None
+            self.count += 2
+            opt = self.optimizers()
+            output = self(batch)
+            loss = self.loss(batch, output)
+            opt.zero_grad()
+            loss.backward()
+            opt.step()
```
Comment (carmocca) [inline tests/utilities/test_fetching.py:291]: Why override `training_step` if the test does not call `fit`?

And `automatic_optimization`?

**Review by carmocca — APPROVED** [review]
